pkg/store/gcs.go

```diff
@@ -428,25 +334,61 @@ func (s *gcsSeriesSet) Err() error {
 func (s *GCSStore) blockSeries(ctx context.Context, b *gcsBlock, matchers []labels.Matcher, mint, maxt int64) (storepb.SeriesSet, error) {
 	var (
 		extLset = b.meta.Thanos.Labels
-		indexr  = b.indexReader()
+		indexr  = b.indexReader(ctx)
 		chunkr  = b.chunkReader(ctx)
 	)
 	defer indexr.Close()
 	defer chunkr.Close()
 
 	begin := time.Now()
-	set, err := tsdb.LookupChunkSeries(indexr, nil, matchers...)
+
+	// The postings to preload are registered within the call to PostingsForMatchers,
+	// when it invokes indexr.Postings for each underlying postings list.
+	p, absent, err := tsdb.PostingsForMatchers(indexr, matchers...)
 	if err != nil {
-		return nil, errors.Wrap(err, "get series set")
+		return nil, err
 	}
-	var res []seriesEntry
 
-	for set.Next() {
-		lset, chks, _ := set.At()
+	level.Debug(s.logger).Log("msg", "setup postings", "duration", time.Since(begin))
+	begin = time.Now()
+
+	if err := indexr.preloadPostings(); err != nil {
+		return nil, err
+	}
+
+	level.Debug(s.logger).Log("msg", "preload postings", "duration", time.Since(begin))
+	begin = time.Now()
 
+	var ps []uint64
+	for p.Next() {
+		ps = append(ps, p.At())
+	}
+	if err := p.Err(); err != nil {
+		return nil, err
+	}
+	if err := indexr.preloadSeries(ps); err != nil {
+		return nil, err
+	}
+	level.Debug(s.logger).Log("msg", "preload series", "count", len(ps), "duration", time.Since(begin))
+
+	var (
+		res  []seriesEntry
+		lset labels.Labels
+		chks []chunks.Meta
+	)
+Outer:
+	for _, id := range ps {
+		if err := indexr.Series(id, &lset, &chks); err != nil {
+			return nil, err
+		}
+		for _, l := range absent {
+			if lset.Get(l) != "" {
+				continue Outer
```

Comment:
Can you comment what is happening here? Why if absent label has some value we ... continue `for _, id := range ps {` loop?